[experiment] Unboxed representation of optionals, and unification with nullables.

cristianoc opened the pull request:
More details in https://github.com/BuckleScript/bucklescript/issues/2801

bobzhang commented:
Here are some of my thoughts of the encoding:

`'a option` is more precise actually in JS typing world (TS/Flow), there is no such counter-part since it can not encoding (Some None)

Predefined marker:
```
let nullHeader = [ |  |] 
let undefinedHeader = [| |]
```
we make use of the fact that js object address can be  uniquely identified so that we can use it to tell it is a boxed null or undefined

Semantics of `Some x`:
```js
function box(x){
   if (x === null) {
         return [nullHeader, 0] 
   } else if (x === undefined) {
        return [undefinedHeader,0]
  } else if (x [0] === nullHeader) {
      return [nullHeader, x[1] + 1]
  } else if (x [0] === undefinedHeader) {
      return [undefinedHeader, x[1] + 1]
  } else return x
}
```

```js
function isNone(x){
   if (x === null || x === undefined ) {
      return true    
   }
  return false
}
```

```js
function unbox(x){
  if( x [0] === nullHeader) {
         if (x[1] === 0) {
          return null
        }
       else return [nullHeader,  x[1] - 1]
  }
  if (x[0] === undefinedHeader){
    if(x[1] === 1) {
        return undefined
    }
   else return [undefinedHeader, x[1] - 1]
  }
}
```

The downside of using `header` is that it requires more work to serialize `Some None`.

`'a option` can be used to model JS ffi, since in JS world, its value can be `null`, `undefined` or regular value so we can remove Js.Nullable

chenglou commented:
Currently progress:

```
None -> undefined
a == None -> a == undefined
a === None -> a === undefined
| None -> a == undefined

Some(1) -> 1
Some(Some(1)) -> 1
Some(None) -> undefinedHeader (one pre-declared [])
Some(Some(None)) -> [undefinedHeader, 2]
```

chenglou commented:
> The downside of using header is that it requires more work to serialize `Some None`.

That's a good point...

chenglou commented:
Ok, to sum up:

- the `=` and `==` semantics change for this particular case
- serialization is troublesome, though less so than today since 90% of the time you won't need a converter anyway (though what exactly are we gonna do about it?)

Anything else?

chenglou commented:
@cristianoc can you rebase please? We should evaluate this more seriously now; otherwise this is gonna drag on for a while

cristianoc commented:
@chenglou  rebased -- notice one needs to generate artifacts (and recompile ocaml as it's been modified) before the tests run. I'm not checking it the artifacts as there's no way back from that.

bobzhang commented:
sorry that I am in the middle of something else, I will do a code review by the end of this week.

bobzhang commented on jscomp/core/js_exp_make.ml:
```diff
@@ -172,6 +178,10 @@ let make_block ?comment tag tag_info es mutable_flag : t =
   in
+  match tag_info, es with
+  | Blk_constructor ("Some", 1), [e] ->
+    make_box e
```
This seems not precise enough

bobzhang commented on vendor/ocaml/bytecomp/matching.ml:
```diff
@@ -1349,7 +1349,11 @@ let make_constr_matching p def ctx = function
+        | Cstr_block 0 when cstr.cstr_name = "Some" && cstr.cstr_arity = 1 ->
+            (* Format.eprintf "XXX Found Some @."; *)
+            (Lprim(Pfield (0, Fld_record "Some"), [arg], p.pat_loc), Alias) :: argl
+            (* make_field_args p.pat_loc Alias arg 0 (cstr.cstr_arity - 1) argl *)
```
Shall we introduce a new primitive here, or `Fld_unbox`

bobzhang commented on vendor/ocaml/bytecomp/translcore.ml:
```diff
@@ -532,10 +532,18 @@ let transl_prim loc prim args =
-        intcomp
+      begin
+        match intcomp with
+        | Pintcomp (c, _) -> Pintcomp (c, Cmp_opt) (* TODO: specialize to optionals *)
```
this seems to be a regression?

bobzhang commented on vendor/ocaml/bytecomp/matching.ml:
```diff
@@ -2439,7 +2443,11 @@ let combine_constructor loc arg ex_pat cstr partial ctx def
+              if cstr.cstr_name = "None" || cstr.cstr_name = "Some" then
+                let none = Lconst(Const_pointer (0, (Lambda.Pt_constructor "None"))) in
+                Lifthenelse(Lprim (Pintcomp (Cneq, Cmp_opt), [arg; none], loc), act2, act1)
+              else
```
can we introduce a specialized instruction here?

cristianoc commented:
Some offline discussion: try to do something smaller first, where only one of nullable/undefined is supported.
This eliminates the main complexities of having to deal with two kinds of equalities.
In particular, supporting only one value avoids this problematic semantic question, which is not easy to address without bringing along all sorts of compromises:
```reason
  x === None

  (is equivalent to ???)

  switch x { | None => true | Some(_) => false}
```

cristianoc commented:
@bobzhang this is ready for another round of review.
The main difference now is that we only support `undefined` for boxing, and not `null`, with the relevant simplifications that this brings.
There's some more tightening to do to the logic which recognizes optionals from the context where Some/None appear, but I'll leave this for later as it's orthogonal to the rest.

bobzhang commented on jscomp/core/js_exp_make.ml:
```diff
@@ -155,6 +155,12 @@ let merge_outer_comment comment (e : t )  =
                 = Some (comment ^ sep ^ s)} 
 
+let make_box_opt e =
+  runtime_call Js_runtime_modules.js_primitive "box_optional" [e] 
```
call it `some`?

bobzhang commented on jscomp/core/js_exp_make.ml:
```diff
@@ -155,6 +155,12 @@ let merge_outer_comment comment (e : t )  =
+  runtime_call Js_runtime_modules.js_primitive "box_optional" [e] 
+
+let make_unbox_opt e =
+  runtime_call Js_runtime_modules.js_primitive "unbox_optional" [e] 
```
`valFromOption`  cc @chenglou

bobzhang commented on jscomp/core/js_exp_make.ml:
```diff
@@ -87,11 +87,11 @@ let var ?comment  id  : t =
-let nil : t = 
-    {expression_desc = Null ; comment = None}
+let nil ?comment () : t = 
+    {expression_desc = Null ; comment}
```
why do we need change here?

bobzhang commented on jscomp/core/js_exp_make.ml:
```diff
@@ -172,6 +178,10 @@ let make_block ?comment tag tag_info es mutable_flag : t =
   in
+  match tag_info, es with
+  | Blk_constructor ("Some", 1, 1), [e] ->
+    make_box_opt e
```
let's add a branch `Blk_some`?

bobzhang commented on jscomp/core/js_fold_basic.ml:
```diff
@@ -49,7 +49,7 @@ let needBlockRuntimeInReleaseMode (tag : J.expression) (tag_info : J.tag_info) =
   | Blk_tuple 
   | Blk_array -> false   
-  | Blk_constructor (_, 1)      
+  | Blk_constructor (_, 1, _)      
```
Blk_some seems  need runtime support for most cases (also need check debug mode)

bobzhang commented on jscomp/core/js_of_lam_option.ml:
```diff
@@ -53,13 +53,15 @@ let get_default_undefined
-    else
+    else *)
       map Runtime_maybe_unwrapped (E.runtime_call Js_runtime_modules.js_primitive "option_get" [arg])
 
```
```ocaml
external f : ?x : int -> ?y:int -> _ = "" [@@bs.obj];;
f ~x:3 ~y:2
```
need make sure this is compiled correctly

bobzhang commented on jscomp/core/js_of_lam_option.ml:
```diff
@@ -70,12 +72,11 @@ let get_default_undefined
-  {expression_desc = Caml_block ( [x], Immutable, E.zero_int_literal , Blk_constructor ("Some",1) );
-   comment = None}
+let some x : J.expression =
+  E.make_box_opt x
```
maybe we should add a primitive to js ast, so it is easy to cancel box/unbox

bobzhang commented on jscomp/core/lam.ml:
```diff
@@ -1597,6 +1598,8 @@ let lam_prim ~primitive:( p : Lambda.primitive) ~args loc : t =
   | Pmakeblock (tag,info, mutable_flag)
     -> prim ~primitive:(Pmakeblock (tag,info,mutable_flag)) ~args loc
+  | Pfield (id, Fld_record "Some") ->
+    prim ~primitive:Punboxopt ~args loc
```
let's add Fld_val_of_option

bobzhang commented on jscomp/core/lam.ml:
```diff
@@ -1909,7 +1912,7 @@ let convert exports lam : _ * _  =
         | "#makemutablelist" ->
-          Pmakeblock(0,Lambda.Blk_constructor("::",1),Mutable)
+          Pmakeblock(0,Lambda.Blk_constructor("::",1,1),Mutable)
         | "#setfield1" ->
```
per suggestion above, this should not be impacted

bobzhang commented on jscomp/runtime/caml_obj.ml:
```diff
@@ -167,11 +167,11 @@ let unsafe_js_compare x y =
-  if b == (Obj.repr Js.undefined) then 1 else
-  let a_type = Js.typeof a in 
+   if a == (Obj.repr Js.null) then -1 else
+   if b == (Obj.repr Js.null) then 1 else
```
let's make diff minimal, this seems reduandant?

bobzhang commented on jscomp/runtime/caml_obj.ml:
```diff
@@ -269,21 +269,20 @@ type eq = Obj.t -> Obj.t -> bool
     if a_type = "string"
     ||  a_type = "number"
     ||  a_type = "boolean"
-    ||  a_type = "undefined"
```
shall we add a test case `Some None = Some None`

bobzhang commented on vendor/ocaml/bytecomp/lambda.mli:
```diff
@@ -31,7 +31,7 @@ type loc_kind =
 type tag_info = 
-  | Blk_constructor of string * int (* Number of non-const constructors*)
+  | Blk_constructor of string * int * int (* Number of non-const and const  constructors *)
   | Blk_tuple
```
Blk_some

bobzhang commented on vendor/ocaml/bytecomp/matching.ml:
```diff
@@ -2439,7 +2443,11 @@ let combine_constructor loc arg ex_pat cstr partial ctx def
+              if cstr.cstr_name = "None" || cstr.cstr_name = "Some" then
+                let none = Lconst(Const_pointer (0, (Lambda.Pt_constructor "None"))) in
+                Lifthenelse(Lprim (Pintcomp Cneq, [arg; none], loc), act2, act1)
+              else
```
so the pattern match over optional happens in two places?

cristianoc replied:
yes

bobzhang commented on vendor/ocaml/bytecomp/translcore.ml:
```diff
@@ -945,7 +945,7 @@ and transl_exp0 e =
             ))
       | Cstr_block n ->
-          let tag_info = (Lambda.Blk_constructor (cstr.cstr_name, cstr.cstr_nonconsts)) in
+          let tag_info = (Lambda.Blk_constructor (cstr.cstr_name, cstr.cstr_nonconsts, cstr.cstr_consts)) in
```
we can do type based optimization here?

cristianoc replied:
Yes this would be a good place to try.
```ml
    cstr_res: type_expr;                (* Type of the result *)

```

cristianoc commented:
@bobzhang I think I addressed all the comments.
If things look OK, I can create a separate PR with only the instrumentations for ocaml.
Separately afterwards, we could investigate type-based optimization, with enough tests to go with it.

cristianoc commented:
@bobzhang the PR with only the ocaml instrumentation part is here: https://github.com/BuckleScript/bucklescript/pull/2869

cristianoc commented:
This is now split into multiple parts:

https://github.com/BuckleScript/bucklescript/pull/2882
https://github.com/BuckleScript/bucklescript/pull/2885
https://github.com/BuckleScript/bucklescript/pull/2886
https://github.com/BuckleScript/bucklescript/pull/2889
https://github.com/BuckleScript/bucklescript/pull/2890

cristianoc commented:
Done separately.
